### PR TITLE
fix: gracefully handle process stdout/stderr EPIPE error

### LINKED ISSUE #186714: Error: write EPIPE endless loop
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.79.2
- OS Version: Ubuntu 23.04

### Steps to Reproduce:

It is kind of difficult to reproduce, but I sometimes encounter the issue during testing vscode with @vscode/test-electron

1. Launch VSCode using `@vscode/test-electron` 
2. When pressing ctrl+c the vscode child process should get disposed but sometimes ends up in a bad state where it is not quitting and consuming lots of cpu
3. Open the vscode logs, specifically 

- `main.log`
- `main.1.log`
- `main.2.log`
- `main.3.log`
- `main.4.log`
- `main.5.log`
- `main.6.log`

All of these log files contain ~50000 lines of the same output:

```
2023-06-30 07:47:13.051 [error] Error: write EPIPE
    at afterWriteDispatched (node:internal/stream_base_commons:160:15)
    at writeGeneric (node:internal/stream_base_commons:151:3)
    at Socket._writeGeneric (node:net:874:11)
    at Socket._write (node:net:886:8)
    at writeOrBuffer (node:internal/streams/writable:391:12)
    at _write (node:internal/streams/writable:332:10)
    at Writable.write (node:internal/streams/writable:336:10)
    at console.value (node:internal/console/constructor:299:16)
    at console.warn (node:internal/console/constructor:381:26)
    at s.send (node:electron/js2c/browser_init:2:86337)
    at _.send (node:electron/js2c/browser_init:2:71895)
    at bt.send (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3814)
    at bt.sendWhenReady (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3536)
    at at.sendToFocused (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:47075)
    at e.CodeApplication.v (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:54819)
    at E.a (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52424)
    at E.onUnexpectedError (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:249)
    at O (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:603)
    at process.<anonymous> (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52489)
    at process.emit (node:events:525:35)
    at process._fatalException (node:internal/process/execution:167:25)
2023-06-30 07:47:13.052 [error] [uncaught exception in main]: Error: write EPIPE
2023-06-30 07:47:13.052 [error] Error: write EPIPE
    at afterWriteDispatched (node:internal/stream_base_commons:160:15)
    at writeGeneric (node:internal/stream_base_commons:151:3)
    at Socket._writeGeneric (node:net:874:11)
    at Socket._write (node:net:886:8)
    at writeOrBuffer (node:internal/streams/writable:391:12)
    at _write (node:internal/streams/writable:332:10)
    at Writable.write (node:internal/streams/writable:336:10)
    at console.value (node:internal/console/constructor:299:16)
    at console.warn (node:internal/console/constructor:381:26)
    at s.send (node:electron/js2c/browser_init:2:86337)
    at _.send (node:electron/js2c/browser_init:2:71895)
    at bt.send (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3814)
    at bt.sendWhenReady (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3536)
    at at.sendToFocused (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:47075)
    at e.CodeApplication.v (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:54819)
    at E.a (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52424)
    at E.onUnexpectedError (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:249)
    at O (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:603)
    at process.<anonymous> (vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52489)
    at process.emit (node:events:525:35)
    at process._fatalException (node:internal/process/execution:167:25)
2023-06-30 07:47:13.052 [error] [uncaught exception in main]: Error: write EPIPE

[a lof of more of the same error]
```

### Reason
I believe the EPIPE error is happening because `console.log`/`console.warn`/`console.error` is writing to a closed stdout or a closed stderr stream. This causes an uncaught exception which is logged to the console and that is where the endless loop comes from.

### Related issue
https://github.com/microsoft/vscode/issues/162295

### PR DESCRIPTION
**Pull request by SimonSiefke**
fixes https://github.com/microsoft/vscode/issues/186714

This handles stdout/stderr EPIPE errors more gracefully: 

The idea is to attach an error listener to process.stdout/process.stderr and only logging the error only to the bufferLogger and not to the console.

### Before

```
// main.log
2023-06-30 07:47:13.051 [error] Error: write EPIPE
    at afterWriteDispatched (node:internal/stream_base_commons:160:15)
    at writeGeneric (node:internal/stream_base_commons:151:3)
    at Socket._writeGeneric (node:net:874:11)
    at Socket._write (node:net:886:8)
    at writeOrBuffer (node:internal/streams/writable:391:12)
    at _write (node:internal/streams/writable:332:10)
    at Writable.write (node:internal/streams/writable:336:10)
    at console.value (node:internal/console/constructor:299:16)
    at console.warn (node:internal/console/constructor:381:26)
    at s.send (node:electron/js2c/browser_init:2:86337)
    at _.send (node:electron/js2c/browser_init:2:71895)
    at bt.send (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3814)
    at bt.sendWhenReady (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3536)
    at at.sendToFocused (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:47075)
    at e.CodeApplication.v (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:54819)
    at E.a (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52424)
    at E.onUnexpectedError (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:249)
    at O (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:603)
    at process.<anonymous> (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52489)
    at process.emit (node:events:525:35)
    at process._fatalException (node:internal/process/execution:167:25)
2023-06-30 07:47:13.052 [error] [uncaught exception in main]: Error: write EPIPE
2023-06-30 07:47:13.052 [error] Error: write EPIPE
    at afterWriteDispatched (node:internal/stream_base_commons:160:15)
    at writeGeneric (node:internal/stream_base_commons:151:3)
    at Socket._writeGeneric (node:net:874:11)
    at Socket._write (node:net:886:8)
    at writeOrBuffer (node:internal/streams/writable:391:12)
    at _write (node:internal/streams/writable:332:10)
    at Writable.write (node:internal/streams/writable:336:10)
    at console.value (node:internal/console/constructor:299:16)
    at console.warn (node:internal/console/constructor:381:26)
    at s.send (node:electron/js2c/browser_init:2:86337)
    at _.send (node:electron/js2c/browser_init:2:71895)
    at bt.send (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3814)
    at bt.sendWhenReady (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3536)
    at at.sendToFocused (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:47075)
    at e.CodeApplication.v (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:54819)
    at E.a (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52424)
    at E.onUnexpectedError (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:249)
    at O (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:603)
    at process.<anonymous> (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52489)
    at process.emit (node:events:525:35)
    at process._fatalException (node:internal/process/execution:167:25)
2023-06-30 07:47:13.052 [error] [uncaught exception in main]: Error: write EPIPE
2023-06-30 07:47:13.052 [error] Error: write EPIPE
    at afterWriteDispatched (node:internal/stream_base_commons:160:15)
    at writeGeneric (node:internal/stream_base_commons:151:3)
    at Socket._writeGeneric (node:net:874:11)
    at Socket._write (node:net:886:8)
    at writeOrBuffer (node:internal/streams/writable:391:12)
    at _write (node:internal/streams/writable:332:10)
    at Writable.write (node:internal/streams/writable:336:10)
    at console.value (node:internal/console/constructor:299:16)
    at console.warn (node:internal/console/constructor:381:26)
    at s.send (node:electron/js2c/browser_init:2:86337)
    at _.send (node:electron/js2c/browser_init:2:71895)
    at bt.send (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3814)
    at bt.sendWhenReady (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3536)
    at at.sendToFocused (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:47075)
    at e.CodeApplication.v (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:54819)
    at E.a (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52424)
    at E.onUnexpectedError (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:249)
    at O (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:603)
    at process.<anonymous> (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52489)
    at process.emit (node:events:525:35)
    at process._fatalException (node:internal/process/execution:167:25)
2023-06-30 07:47:13.052 [error] [uncaught exception in main]: Error: write EPIPE
2023-06-30 07:47:13.052 [error] Error: write EPIPE
    at afterWriteDispatched (node:internal/stream_base_commons:160:15)
    at writeGeneric (node:internal/stream_base_commons:151:3)
    at Socket._writeGeneric (node:net:874:11)
    at Socket._write (node:net:886:8)
    at writeOrBuffer (node:internal/streams/writable:391:12)
    at _write (node:internal/streams/writable:332:10)
    at Writable.write (node:internal/streams/writable:336:10)
    at console.value (node:internal/console/constructor:299:16)
    at console.warn (node:internal/console/constructor:381:26)
    at s.send (node:electron/js2c/browser_init:2:86337)
    at _.send (node:electron/js2c/browser_init:2:71895)
    at bt.send (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3814)
    at bt.sendWhenReady (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3536)
    at at.sendToFocused (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:47075)
    at e.CodeApplication.v (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:54819)
    at E.a (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52424)
    at E.onUnexpectedError (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:249)
    at O (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:603)
    at process.<anonymous> (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52489)
    at process.emit (node:events:525:35)
    at process._fatalException (node:internal/process/execution:167:25)
2023-06-30 07:47:13.052 [error] [uncaught exception in main]: Error: write EPIPE
2023-06-30 07:47:13.052 [error] Error: write EPIPE
    at afterWriteDispatched (node:internal/stream_base_commons:160:15)
    at writeGeneric (node:internal/stream_base_commons:151:3)
    at Socket._writeGeneric (node:net:874:11)
    at Socket._write (node:net:886:8)
    at writeOrBuffer (node:internal/streams/writable:391:12)
    at _write (node:internal/streams/writable:332:10)
    at Writable.write (node:internal/streams/writable:336:10)
    at console.value (node:internal/console/constructor:299:16)
    at console.warn (node:internal/console/constructor:381:26)
    at s.send (node:electron/js2c/browser_init:2:86337)
    at _.send (node:electron/js2c/browser_init:2:71895)
    at bt.send (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3814)
    at bt.sendWhenReady (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:3536)
    at at.sendToFocused (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:47075)
    at e.CodeApplication.v (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:54819)
    at E.a (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52424)
    at E.onUnexpectedError (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:249)
    at O (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:7:603)
    at process.<anonymous> (.vscode-test/vscode-linux-x64-1.79.2/resources/app/out/vs/code/electron-main/main.js:89:52489)
    at process.emit (node:events:525:35)
    at process._fatalException (node:internal/process/execution:167:25)
```

### After

```
// main.log
2023-06-30 11:26:22.886 [info] Extension host with pid 379203 exited with code: 0, signal: unknown.
2023-06-30 11:26:22.888 [error] stdout error in main: Error: write EPIPE
```

### CODE REVIEW

**Comment by connor4312** [comment]
For consistency, I would prefer if we handled the error in uncaughtException like we do here:

https://github.com/microsoft/vscode/blob/69cd1439e59e494827f7b20e82f28c88c0319d1e/src/vs/code/node/cliProcessMain.ts#L243-L245

**Review by connor4312 — APPROVED** [review]

**Comment by connor4312** [comment]
Thanks for the PR!

File: src/vs/code/electron-main/app.ts [hunk 1]
```diff
@@ -336,7 +336,11 @@ export class CodeApplication extends Disposable {
 
 		// We handle uncaught exceptions here to prevent electron from opening a dialog to the user
 		setUnexpectedErrorHandler(error => this.onUnexpectedError(error));
-		process.on('uncaughtException', error => onUnexpectedError(error));
+		process.on('uncaughtException', error => {
+			if (!isSigPipeError(error)) {
```
Comment (bpasero) [inline src/vs/code/electron-main/app.ts:340]: Should we move this into the `onUnexpectedError`?

Comment (connor4312) [reply src/vs/code/electron-main/app.ts:340]: I thought about that initially when doing the first pass of this, but opted not to for two reasons:

- it's called from many more places where the loop only happens in `process.on('uncaughtException`
- we don't know if `errorHandler.onUnexpectedError` will actually log to the console and cause a loop

...but that may be a better idea considering the virality of this check. What do you think?

**Review by rzhao271 — APPROVED** [review]

File: src/vs/code/electron-main/app.ts (hunk 1 above)
Comment (bpasero) [reply src/vs/code/electron-main/app.ts:340]: Ah ok, I think its fine then.

**Pull request merged**